server/src/main/java/org/apache/iotdb/db/query/udf/core/transformer/BinaryTransformer.java

```diff
@@ -0,0 +1,147 @@
+      currentNull = true;
+    } else {
+      switch (getDataType()) {
+        case DOUBLE:
+          cachedDouble =
+              evaluateDouble(
+                  castCurrentValueToDoubleOperand(leftPointReader),
+                  castCurrentValueToDoubleOperand(rightPointReader));
+          break;
+        case BOOLEAN:
+          cachedBoolean =
+              evaluateBoolean(
+                  castCurrentValueToDoubleOperand(leftPointReader),
+                  castCurrentValueToDoubleOperand(rightPointReader));
+      }
+    }
+    leftPointReader.readyForNext();
+    rightPointReader.readyForNext();
+    return true;
+  }
+
+  /**
+   * finds the smallest, unconsumed timestamp that exists in both {@code leftPointReader} and {@code
+   * rightPointReader} and then caches the timestamp in {@code cachedTime}.
+   *
+   * @return true if there has a timestamp that meets the requirements
+   */
+  private boolean cacheTime() throws IOException, QueryProcessException {
+    if (leftPointReader.isConstantPointReader() && rightPointReader.isConstantPointReader()) {
+      return true;
+    }
+    if (leftPointReader.isConstantPointReader()) {
+      cachedTime = rightPointReader.currentTime();
+      return true;
+    }
+    if (rightPointReader.isConstantPointReader()) {
+      cachedTime = leftPointReader.currentTime();
+      return true;
+    }
+
+    long leftTime = leftPointReader.currentTime();
+    long rightTime = rightPointReader.currentTime();
+
+    while (leftTime != rightTime) {
+      if (leftTime < rightTime) {
+        leftPointReader.readyForNext();
+        if (!leftPointReader.next()) {
+          return false;
+        }
+        leftTime = leftPointReader.currentTime();
+      } else {
+        rightPointReader.readyForNext();
+        if (!rightPointReader.next()) {
+          return false;
+        }
+        rightTime = rightPointReader.currentTime();
+      }
+    }
+
+    // leftTime == rightTime
+    cachedTime = leftTime;
+    return true;
+  }
+
+  protected double evaluateDouble(double leftOperand, double rightOperand) {
+    return 0.0;
+  }
+
+  protected boolean evaluateBoolean(double leftOperand, double rightOperand) {
+    return false;
+  }
+
+  private static double castCurrentValueToDoubleOperand(LayerPointReader layerPointReader)
+      throws IOException, QueryProcessException {
+    switch (layerPointReader.getDataType()) {
+      case INT32:
+        return layerPointReader.currentInt();
+      case INT64:
+        return layerPointReader.currentLong();
+      case FLOAT:
+        return layerPointReader.currentFloat();
+      case DOUBLE:
+        return layerPointReader.currentDouble();
+      case BOOLEAN:
+        return layerPointReader.currentBoolean() ? 1.0 : 0.0;
+      default:
+        throw new QueryProcessException(
+            "Unsupported data type: " + layerPointReader.getDataType().toString());
+    }
+  }
+
+  @Override
+  public TSDataType getDataType() {
```

Comment:
I think this one should also be abstract as the sub-classes may return `DOUBLE` or `BOOLEAN`.